lib/fxpay/adapter.js

```diff
@@ -0,0 +1,166 @@
+(function() {
+  'use strict';
+
+  var exports = window.fxpay.utils.namespace('fxpay.adapter');
+
+  var API = require('fxpay/api').API;
+  var products = require('fxpay/products');
+  var receipts = require('fxpay/receipts');
+  var settings = require('fxpay/settings');
+  var utils = require('fxpay/utils');
+
+
+  function FxInappAdapter() {
+    //
+    // Adapter for Firefox Marketplace in-app products.
+    //
+    // Create your own object with these same methods to adapt
+    // another payment fulfillment system. This adapter was
+    // designed to accomodate Firefox Marketplace app purchases
+    // which differ slightly from in-app products.
+    //
+    settings.log.info('using Firefox Marketplace In-App adapter');
+    this.api = new API(settings.apiUrlBase);
+  }
+
+  FxInappAdapter.prototype.init = function(callback) {
+    //
+    // Initialize the payment system.
+    //
+    // This is called when the fxpay library itself is initialized.
+    // This is a chance to restore purchases or anything that should
+    // happen when an app starts up.
+    //
+    // This must execute callback(error) when finished.
+    // The error parameter should be null when there are no errors.
+    //
+    if (!settings.hasAddReceipt && !settings.localStorage) {
+      settings.log.error('no way to store receipts on this platform');
+      return callback('PAY_PLATFORM_UNAVAILABLE');
+    }
+    if (settings.appSelf &&
+        settings.window.location.href.indexOf('app://') === 0 &&
+        !settings.appSelf.manifest.origin) {
+      settings.log.error('packaged app did not define an origin so ' +
+                         'we have no key to look up products');
+      return callback('UNDEFINED_APP_ORIGIN');
+    }
+    var numReceipts = 0;
+    var receipt;
+    var allReceipts = receipts.all();
+    for (var i = 0; i < allReceipts.length; i++) {
+      receipt = allReceipts[i];
+      settings.log.info('Installed receipt: ' + receipt);
+      numReceipts++;
+      receipts.verify(receipt, settings.onrestore);
+    }
+    settings.log.info('Number of receipts installed: ' + numReceipts);
+    callback();
+  };
+
+  FxInappAdapter.prototype.startTransaction = function(opt, callback) {
+    //
+    // Start a transaction.
+    //
+    // The `opt` object contains the following parameters:
+    //
+    // - productId: the ID of the product purchased.
+    //
+    // When finished, execute callback(error, transactionData).
```

Comment:
nice docs - thinking we could look at generating some code docs with these kind of comments. E.g. yuidoc or similar.